py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,427 @@
+                time_to = delta_time(end_dt)
+                time_to_formatted = '{} {}'.format(self._format_timedelta(index, time_to),
+                                                   self.event_time_suffix)
+            else:
+                time_to_formatted = self._format_timedelta(index, time_to)
+
+            self._check_button_open(index, url)
+
+            self.py3.threshold_get_color(index+1)
+
+            curr_event_format = self._check_button_toggle(index)
+
+            event_formatted = self.py3.safe_format(
+                curr_event_format,
+                {'summary': summary,
+                 'location': location,
+                 'description': description,
+                 'start_time': start_time_str,
+                 'end_time': end_time_str,
+                 'start_date': start_date_str,
+                 'end_date': end_date_str,
+                 'time_to': time_to_formatted})
+
+            self.py3.composite_update(event_formatted, {'index': index})
+            responses.append(event_formatted)
+
+            index += 1
+
+        format_separator = self.py3.safe_format(self.format_separator)
+        responses = self.py3.composite_join(format_separator, responses)
+
+        return {'events': responses}
+
+    def google_calendar(self):
+        """
+        The method that outputs the response.
+
+        First, we check credential authorization. If no authorization, we display an error message,
+        and try authorizing again in 5 seconds.
+
+        Otherwise, we fetch the events, build the response, and output the resulting composite.
+        """
+        if not self.is_authorized:
+            self.is_authorized = self._authorize_credentials()
+
+            error_msg = self.py3.safe_format(
+                self.format_error,
+                {'error':
+                 'ServerNotFoundError: Check your internet connection!'})
+
+            return {'cached_until': self.py3.time_in(5),
+                    'full_text': error_msg,
+                    'color': '#FF0000'}
+        else:
+            if not self.no_update:
+                self.events = self._get_events()
+
+            return {
+                'cached_until': self.py3.time_in(self.cache_timeout),
+                'composite': self.py3.safe_format(self.format, self._build_response())
+            }
+
+    def on_click(self, event):
+        if self.is_authorized and self.events is not None:
+            """
+            If a button is clicked, we stop updating self.events until the
+            button event is handled. This prevents errors when the on_click()
+            method is called during an event update (using _get_events()), which
+            can cause a crash in the module and/or py3status.
+            """
+            try:
+                self.no_update = True
+
+                self.button = event['button']
+
+                self.button_index = event['index']
+                if self.button == self.button_toggle:
```

Comment:
We might want `self.button_refresh` here to allow users to forcefully refresh if they want to.
We might want `self.py3.prevent_refresh()` at bottom... so scrolling away does not make `google_calendar` update rapidly for little or no gain.

Reply:
Hmm, the problem with this is it reintroduces a bug I had earlier. To implement this, the module would have to fetch the events using the API on every refresh. The problem is, if this is done quickly, py3status can crash. 

It's hard to debug because it only happens when the module (and its events) are refreshed with a bunch of quick button presses. For example, if I set button 2 to refresh the module and all the events, if I press button 2 over and over very quickly, the module will eventually crash. There seems to be some problem when some events are being fetched, and the process is interrupted to call `_get_events()` again. If this is done fast enough, the `events` dict contains garbage, and that takes down the module. Unfortunately, there's nothing in the py3status log that's helpful in figuring it out.

That's why I have that `self.no_update` boolean. I use it to make sure we don't fetch events until a button event is handled, so that an event update cannot be interrupted. This prevents the crash from happening.